pkg/policy/eval.go

```diff
@@ -0,0 +1,69 @@
+//
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policy
+
+import (
+	"context"
+	"fmt"
+
+	"cuelang.org/go/cue/cuecontext"
+	cuejson "cuelang.org/go/encoding/json"
+
+	"knative.dev/pkg/logging"
+)
+
+// EvaluatePolicyAgainstJson is used to run a policy engine against JSON bytes.
+// These bytes can be for example Attestations, or ClusterImagePolicy result
+// types.
+// name - which attestation are we evaluating
+// policyType - cue|rego
+// policyBody - String representing either cue or rego language
+// jsonBytes - Bytes to evaluate against the policyBody in the given language
+func EvaluatePolicyAgainstJSON(ctx context.Context, name, policyType string, policyBody string, jsonBytes []byte) error {
+	logging.FromContext(ctx).Debugf("Evaluating JSON: %s against policy: %s", string(jsonBytes), policyBody)
+	switch policyType {
+	case "cue":
+		cueValidationErr := evaluateCue(ctx, jsonBytes, policyBody)
+		if cueValidationErr != nil {
+			return fmt.Errorf("failed evaluating cue policy for %s : %s", name, cueValidationErr.Error()) // nolint
+		}
+	case "rego":
+		regoValidationErr := evaluateRego(ctx, jsonBytes, policyBody)
+		if regoValidationErr != nil {
+			return fmt.Errorf("failed evaluating rego policy for type %s", name)
+		}
+	default:
+		return fmt.Errorf("sorry Type %s is not supported yet", policyType)
+	}
+	return nil
+}
+
+// evaluateCue evaluates a cue policy `evaluator` against `attestation`
+func evaluateCue(ctx context.Context, attestation []byte, evaluator string) error {
+	logging.FromContext(ctx).Infof("Evaluating attestation: %s", string(attestation))
+	cueCtx := cuecontext.New()
+	v := cueCtx.CompileString(evaluator)
+	return cuejson.Validate(attestation, v)
```

Comment:
As discussed, we have to use a different set of functions here, as cuejson.Validate expects the CUE to be self-contained and applied as a schema to the JSON.

Reply:
We can improve this function in a different PR.